### PR TITLE
Add structured-pruned YOLOv8 KD compatibility

### PR DESCRIPTION
**Pull request by minuex**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds structured-pruned YOLOv8 model reconstruction and official knowledge-distillation training compatibility, while applying broad Python style and documentation cleanup.

### 📊 Key Changes
- Adds `C2fPruned`, `SPPFPruned`, and `DetectPruned` modules that rebuild channel widths from saved pruning masks while preserving the official detection-head output contract.
- Adds `DetectionModelPruned` and `parse_model_pruned` to reconstruct pruned YOLOv8 graphs, including dataset class-count overrides and checkpoint mask preservation.
- Updates the detection trainer to select the pruned model implementation when loading checkpoints containing pruning metadata.
- Adds compatibility documentation in `PRUNED_KD.md` and a test covering pruned detection-head outputs and distillation-layer discovery.
- Refactors numerous code examples and implementation details, including dictionary literals, imports, no-op method bodies, exception re-raising, and iteration patterns.

### 🎯 Purpose & Impact
Structured-pruned YOLOv8 checkpoints containing `maskbndict` can be rebuilt through the detection training flow and used with the existing distillation model interface, including fine-tuning with a different dataset class count. The remaining edits are primarily maintenance, style, and documentation updates.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @minuex, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/yolo/detect/train.py [hunk 1]
```diff
@@ -195,9 +196,22 @@ def get_model(self, cfg: str | None = None, weights: str | None = None, verbose:
         Returns:
             (DetectionModel): YOLO detection model.
         """
-        model = self.set_model_names_for_load(
-            DetectionModel(cfg, nc=self.data["nc"], ch=self.data["channels"], verbose=verbose and RANK == -1)
-        )
+        if isinstance(weights, DetectionModelPruned):
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/train.py:199]: ⚠️ **HIGH**: The trainer is supposed to select the pruned implementation for checkpoints containing pruning metadata, but this checks the serialized model's exact Python class instead of the metadata that `load_checkpoint()` attaches at lines 1853-1857. A checkpoint with `maskbndict` on a regular or legacy model class falls through to `DetectionModel`, and `model.load(weights)` cannot rebuild the pruned channel dimensions. Select this path based on a valid `maskbndict` (or a dedicated checkpoint flag), not only `isinstance(weights, DetectionModelPruned)`.

File: ultralytics/nn/tasks_pruned.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        LOGGER.info(f"\n{'':>3}{'from':>20}{'n':>3}{'params':>10}  {'module':<50}{'arguments':<30}")
+    ch = [ch]
+    layers, save, c2 = [], [], ch[-1]  # layers, savelist, ch out
+    # 按照{某一bn层的名字: 该bn层连接的上一(多)bn层的名字}记录
+    # 因为某一层的输出通道哪些保留哪些不保留是由mask来定的,
+    # 某一层的输入通道哪些保留哪些不保留是由上一层的mask来定的
+    current_to_prev = {}
+    # 按照{索引: 本层最后一个bn层的名字}记录
+    idx_to_bn_layer_name = {}
+    for i, (f, n, m, args) in enumerate(d["backbone"] + d["head"]):  # from, number, module, args
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks_pruned.py:123]: ⚠️ **HIGH**: The attached YOLOv8 `model.yaml` stores module names as strings such as `"Conv"`, `"C2f"`, `"SPPF"`, `"nn.Upsample"`, and `"Detect"`, but this parser never resolves `m` or maps those standard names to the pruned classes. The first layer therefore skips every branch and reaches the unsupported-module error, so `DetectionModelPruned` cannot reconstruct the advertised checkpoints. Resolve the module names and map the YOLOv8 modules to `Conv`, `C2fPruned`, `SPPFPruned`, `Concat`, `nn.Upsample`, and `DetectPruned` before dispatching.

File: .github/scripts/create-export-env.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 REPO = Path(__file__).resolve().parents[2]
 sys.path.insert(0, str(REPO))
 
-from ultralytics.engine.exporter import EXPORT_ENVS  # noqa: E402
+from ultralytics.engine.exporter import EXPORT_ENVS
```
Comment (UltralyticsAssistant) [inline .github/scripts/create-export-env.py:16]: 💡 **MEDIUM**: Removing `# noqa: E402` leaves an import after executable `sys.path.insert()` statements. Ruff's default E4 rules include E402, so the lint job will fail on this script. Restore the exemption or restructure the path setup/import.

**Suggested change:**
```suggestion
from ultralytics.engine.exporter import EXPORT_ENVS  # noqa: E402
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the full substantive diff, including checkpoint loading, pruned graph reconstruction, KD integration, and the mechanical refactors. The PR is not ready: the pruned parser cannot consume the normal YOLOv8 YAML representation, metadata-based checkpoint selection is too restrictive, and there are additional safe-load, custom-channel, profiling, and lint failures. The remaining style-only edits appear behavior-preserving.

<details><summary>💬 Posted 6 inline comments</summary>

- ⚠️ **HIGH** `ultralytics/models/yolo/detect/train.py:199` The trainer is supposed to select the pruned implementation for checkpoints containing pruning metadata, but this checks the serialized model's exact Python class instead of the metadata that `load_checkpoint()` attaches at lines 1853-1857. A checkpoint with `maskbndict` on a regular or legacy model class falls through to `DetectionModel`, and `model.load(weights)` cannot rebuild the pruned channel dimensions. Select this path based on a valid `maskbndict` (or a dedicated checkpoint flag), not…
- ⚠️ **HIGH** `ultralytics/nn/tasks_pruned.py:123` The attached YOLOv8 `model.yaml` stores module names as strings such as `"Conv"`, `"C2f"`, `"SPPF"`, `"nn.Upsample"`, and `"Detect"`, but this parser never resolves `m` or maps those standard names to the pruned classes. The first layer therefore skips every branch and reaches the unsupported-module error, so `DetectionModelPruned` cannot reconstruct the advertised checkpoints. Resolve the module names and map the YOLOv8 modules to `Conv`, `C2fPruned`, `SPPFPruned`, `Concat`, `nn.Upsample`, and…
- 💡 **MEDIUM** `.github/scripts/create-export-env.py:16` Removing `# noqa: E402` leaves an import after executable `sys.path.insert()` statements. Ruff's default E4 rules include E402, so the lint job will fail on this script. Restore the exemption or restructure the path setup/import.
- 💡 **MEDIUM** `ultralytics/nn/tasks_pruned.py:18` `ULTRALYTICS_SAFE_LOAD` cannot load checkpoints serialized with this new model class: `_SafeLoad._build()` scans `ultralytics.nn.tasks` and `ultralytics.nn.modules`, but not `ultralytics.nn.tasks_pruned`. Such checkpoints are rejected as unsupported globals before `load_checkpoint()` can preserve their masks. Add `tasks_pruned` to the safe-load allow-list scan or explicitly register `DetectionModelPruned`.
- 💡 **MEDIUM** `ultralytics/nn/tasks_pruned.py:25` Unlike `_initialize_yolo_model()`, this reconstruction never stores the actual input channel count in `self.yaml["channels"]`. `DistillationModel` reads that field and defaults to 3; loading a pruned grayscale or other non-RGB model for KD will then create a 3-channel dummy input against a model built with `ch=1` (or another custom value), failing during hook/projector initialization. Preserve the passed `ch` in the rebuilt YAML.
- 📝 **LOW** `ultralytics/nn/tasks_pruned.py:267` `m` is the module class, while `m_` is the instantiated layer. Assigning `np` to `m` creates a class-level value shared by all layers of that type and leaves the instance without its own parameter count, so later profiling reports the last layer's count for earlier layers. Assign the count to `m_` instead.
</details>

File: ultralytics/nn/tasks_pruned.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+from ultralytics.nn.modules.block_pruned import C2fPruned, SPPFPruned
+from ultralytics.nn.modules.conv import Concat, Conv
+from ultralytics.nn.modules.head_pruned import DetectPruned
+from ultralytics.nn.tasks import DetectionModel
+from ultralytics.utils import LOGGER, colorstr
+from ultralytics.utils.torch_utils import initialize_weights
+
+
+class DetectionModelPruned(DetectionModel):
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks_pruned.py:18]: 💡 **MEDIUM**: `ULTRALYTICS_SAFE_LOAD` cannot load checkpoints serialized with this new model class: `_SafeLoad._build()` scans `ultralytics.nn.tasks` and `ultralytics.nn.modules`, but not `ultralytics.nn.tasks_pruned`. Such checkpoints are rejected as unsupported globals before `load_checkpoint()` can preserve their masks. Add `tasks_pruned` to the safe-load allow-list scan or explicitly register `DetectionModelPruned`.

File: ultralytics/nn/tasks_pruned.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    """Structured-pruned YOLOv8 model using the official DetectionModel contract."""
+
+    def __init__(self, maskbndict, cfg, ch=3, nc=None, verbose=True):
+        """Build the pruned graph from its saved BN masks and model YAML."""
+        nn.Module.__init__(self)
+        self.maskbndict = maskbndict
+        self.yaml = deepcopy(cfg)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks_pruned.py:25]: 💡 **MEDIUM**: Unlike `_initialize_yolo_model()`, this reconstruction never stores the actual input channel count in `self.yaml["channels"]`. `DistillationModel` reads that field and defaults to 3; loading a pruned grayscale or other non-RGB model for KD will then create a 3-channel dummy input against a model built with `ch=1` (or another custom value), failing during hook/projector initialization. Preserve the passed `ch` in the rebuilt YAML.

File: ultralytics/nn/tasks_pruned.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                current_to_prev[cv2x2_out_conv_layer_names[ix]] = cv2x1_out_bn_layer_names[ix]
+                current_to_prev[cv3x2_out_conv_layer_names[ix]] = cv3x1_out_bn_layer_names[ix]
+            # =============================================
+        else:
+            raise ValueError(f"ERROR ❌ module {m} not supported in parse_model.")
+        prev_module = m
+
+        m_ = nn.Sequential(*(m(*args) for _ in range(n))) if n > 1 else m(*args)  # module
+        t = str(m)[8:-2].replace("__main__.", "")  # module type
+        m.np = sum(x.numel() for x in m_.parameters())  # number params
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks_pruned.py:267]: 📝 **LOW**: `m` is the module class, while `m_` is the instantiated layer. Assigning `np` to `m` creates a class-level value shared by all layers of that type and leaves the instance without its own parameter count, so later profiling reports the last layer's count for earlier layers. Assign the count to `m_` instead.

**Suggested change:**
```suggestion
        m_.np = sum(x.numel() for x in m_.parameters())
```

**Comment by glenn-jocher** [comment]
@minuex, please sign the Ultralytics Contributor License Agreement by commenting exactly: `I have read the CLA Document and I sign the CLA`.

**Comment by glenn-jocher** [comment]
Closing because this does not meet the repository acceptance bar. The PR changes 103 files and adds 513 net lines, most of them unrelated mechanical rewrites, while the core feature is not functional: normal YOLOv8 YAML module names are not resolved by the pruned parser, and the trainer selects the pruned path by serialized Python class rather than the pruning metadata it claims to support. It is also conflicting and the CLA remains unsigned. A future contribution would need a signed CLA, an isolated production-relevant change in the existing model/trainer owners, and a minimal real-checkpoint reproduction with no repository-wide cleanup.
